`supply-main-arguments: empty` causes inscrutable `No implementation or override found for pointer` error

This simple program:

```c
int a;

int getService() {
  a = 9;
  return 0;
}

int main(int argc, char *argv[]) {
  return getService();
}
```

Will fail when `supply-main-arguments: empty` is used in a config file:

```
$ cabal run exe:crux-llvm -- --config test2.config test2.c
Resolving dependencies...
Up to date
[Crux] Using pointer width: 64 for file crux-build/crux~test2.bc
[Crux] Simulating function main
[Crux] Attempting to prove verification conditions.
[Crux] *** debug executable: results/test2/debug-9
[Crux] *** break on line: 9
[Crux] Found counterexample for verification goal
[Crux]   test2.c:9:10: error: in main
[Crux]   Failed to load function handle
[Crux]   Details:
[Crux]     No implementation or override found for pointer: "getService"
[Crux]     The given pointer could not be resolved to a callable function
[Crux]     No implementation or override found for pointer: "getService"
[Crux]     Attempting to load callable function with type: i32()
[Crux]       Via pointer: Global symbol "getService" (2, 0x0:[64])
[Crux]     In memory state:
[Crux]       Stack frame main
[Crux]         No writes or allocations
[Crux]       Base memory
[Crux]         Allocations:
[Crux]           StackAlloc 5 0x0:[64] Mutable 8-byte-aligned crux-llvm main(argc, argv) simulation
[Crux]           GlobalAlloc 4 0x4:[64] Mutable 4-byte-aligned [global variable  ] a
[Crux]           GlobalAlloc 3 0x0:[64] Immutable 1-byte-aligned [defined function ] main
[Crux]           GlobalAlloc 2 0x0:[64] Immutable 1-byte-aligned [defined function ] getService
[Crux]           GlobalAlloc 1 0x0:[64] Immutable 1-byte-aligned [external function] llvm.dbg.value
[Crux]         Writes:
[Crux]           Indexed chunk:
[Crux]             4 |->   *(4, 0x0:[64]) := 0
[Crux] Goal status:
[Crux]   Total: 1
[Crux]   Proved: 0
[Crux]   Disproved: 1
[Crux]   Incomplete: 0
[Crux]   Unknown: 0
[Crux] Overall status: Invalid.
```

On the other hand, the same program when `int main() { ... }` is used instead.

--- a/crux-llvm/src/Crux/LLVM/Simulate.hs
+++ b/crux-llvm/src/Crux/LLVM/Simulate.hs
@@ -21,7 +21,7 @@ import qualified Data.List as List
 import Data.Maybe ( fromMaybe )
 import qualified Data.Parameterized.Map as MapF
 import qualified Data.Traversable as T
-import Control.Lens ((&), (%~), (%=), (^.), view)
+import Control.Lens ((&), (%~), (%=), (^.), use, view)
 import Control.Monad.State(liftIO)
 import Data.Text as Text (Text, pack)
 import GHC.Exts ( proxy# )
@@ -200,7 +200,7 @@ setupFileSim halloc llvm_file llvmOpts sym _maybeOnline =
              withPtrWidth ptrW $
                 do registerFunctions llvmOpts (prepLLVMMod prepped) trans (Just fs0)
                    initFSOverride
-                   checkFun llvmOpts trans prepped
+                   checkFun llvmOpts trans memVar
 
 
 data PreppedLLVM sym = PreppedLLVM { prepLLVMMod :: LLVM.Module
@@ -245,9 +245,9 @@ checkFun ::
   Log.SupportsCruxLLVMLogMessage msgs =>
   LLVMOptions ->
   ModuleTranslation arch ->
-  PreppedLLVM sym ->
+  GlobalVar Mem ->
   OverM personality sym LLVM ()
-checkFun llvmOpts trans prepped =
+checkFun llvmOpts trans memVar =
   case Map.lookup (fromString nm) mp of
     Just (_, AnyCFG anyCfg) ->
       case cfgArgTypes anyCfg of
@@ -263,8 +263,6 @@ checkFun llvmOpts trans prepped =
   where
     nm     = entryPoint llvmOpts
     mp     = cfgMap trans
-    memVar = prepMemVar prepped
-    mem    = prepMem prepped
     isMain = nm == "main"
     shouldSupplyMainArguments =
       case supplyMainArguments llvmOpts of
@@ -306,6 +304,10 @@ checkFun llvmOpts trans prepped =
           alignment = memTypeAlign dl tp
           loc       = "crux-llvm main(argc, argv) simulation"
 
+      gs  <- use (stateTree.actFrame.gpGlobals)
+      mem <- case lookupGlobal memVar gs of
+               Just mem -> pure mem
+               Nothing  -> fail "checkFun.checkMainWithArguments: Memory missing from global vars"
       argc <- liftIO $ LLVMPointer <$> natLit sym 0 <*> bvLit sym w (BV.zero w)
       sz   <- liftIO $ bvLit sym PtrWidth $ bytesToBV PtrWidth tp_sz
       (argv, mem') <- liftIO $ doAlloca sym mem sz alignment loc

--- a/crux-llvm/test-data/golden/T816.c
+++ b/crux-llvm/test-data/golden/T816.c
@@ -1,3 +1,16 @@
+#include <stdlib.h>
+
+int a = 42;
+
+void f(int b) {
+  a = b;
+}
+
 int main(int argc, char *argv[]) {
-  return 0;
+  f(argc);
+  if (!a) {
+    return 0;
+  } else {
+    abort();
+  }
 }

